cmip6_downscaling/analysis/metrics.py

```diff
@@ -59,3 +60,154 @@ def calc(obj, compute=False):
+    ds: xr.Dataset, kind_of_day: str = 'wet', threshold: float = 0.01
+):
+    """Contitional probability of a day being the same as the day before.
+    So, given the fact that it was wet yesterday, what is the probability it will be wet today.
+    Or, if it was dry yesterday, what is probability it will be dry today.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        dataset including daily precipitation
+    kind_of_day : str, optional
+        The kind of day you are interested in, by default 'wet'
+    threshold : float, optional
+        Threshold to determine whether day is wet or dry, by default 0.01
+
+    Returns
+    -------
+    xr.Dataset
+        Probability of a day being the same as the day before it
+    """
+    if kind_of_day == 'wet':
+        valid_days = is_wet_day(ds, threshold=threshold)
+    elif kind_of_day == 'dry':
+        valid_days = ~is_wet_day(ds, threshold=threshold)
+    yesterday_valid = valid_days.roll({'time': 1})
+    return valid_days.where(yesterday_valid).mean(dim='time', skipna=True)
+
+
+def probability_wet(ds: xr.Dataset, threshold: float = 0.01):
+    """
+    Unconditional probability of a day being wet
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Daily dataset with precipitation
+    threshold : float, optional
+        Threshold to determine whether day is wet or dry, by default 0.01
+    """
+    return is_wet_day(ds, threshold=threshold).mean(dim='time')
+
+
+def spell_length_stat(timeseries: np.array, method: str = 'mean') -> float:
+    """
+    Statistic about length of a spell of boolean values.
+
+    Parameters
+    ----------
+    timeseries : numpy array
+        Array of booleans denoting a day being (for example) wet or dry.
+    method : str
+        String corresponding to the final statistic you want (mean, std, or percentile)
+    """
+    # add a 0 to help with spells at beginning or end of series
+    timeseries = np.append(timeseries, [0])
+    yesterday = np.roll(timeseries, 1)
+    changes = timeseries - yesterday
+    spell_starts = np.argwhere(changes == 1)
+    spell_ends = np.argwhere(changes == -1)
+    spells = spell_ends - spell_starts
+    if method == 'mean':
+        return np.mean(spells)
+    elif method == 'std':
+        return np.std(spells)
+    elif 'percentile' in method:
+        # parse the percentile
+        percentile = float(method.split('percentile')[1]) / 100
+        return np.quantile(spells, percentile)
+    else:
+        raise NotImplementedError
+
+
+def apply_spell_length(wet_days: xr.Dataset, metric: str = 'mean'):
+    wet_spell_length = xr.apply_ufunc(
+        spell_length_stat, wet_days, metric, input_core_dims=[['time'], []], vectorize=True
+    )
+    return wet_spell_length
+
+
+def monthly_variability(ds, method='sum'):
+    if method == 'sum':
+        return ds.groupby('time.month').sum().std(dim='time')
+    elif method == 'mean':
+        return ds.groupby('time.month').sum().std(dim='time')
```

Comment:
```suggestion
    elif method == 'mean':
        return ds.groupby('time.month').sum().mean(dim='time')
```

see tips above on making this more generic

Reply:
EAGLE EYES! 🦅 👁️ 👁️

Reply:
i like the cleaned up version a lot more